ssh2/session.pyx

```diff
@@ -45,18 +50,49 @@ LIBSSH2_HOSTKEY_TYPE_UNKNOWN = c_ssh2.LIBSSH2_HOSTKEY_TYPE_UNKNOWN
 LIBSSH2_HOSTKEY_TYPE_RSA = c_ssh2.LIBSSH2_HOSTKEY_TYPE_RSA
 LIBSSH2_HOSTKEY_TYPE_DSS = c_ssh2.LIBSSH2_HOSTKEY_TYPE_DSS
 
+cdef ssize_t _send_callback(int fd,
+                            char* buffer,
+                            size_t length,
+                            int flags,
+                            void** abstract) with gil:
+    global errno
+    try:
+        bp = PyBytes_FromStringAndSize(buffer, length)
+        res = (<Session>deref(abstract))._send_callback(fd, bp, flags)
+        return length
+    except Exception as e:
+        return -e.errno
+
+cdef ssize_t _recv_callback(int fd,
+                            char* buffer,
+                            size_t length,
+                            int flags,
+                            void** abstract) with gil:
+    global errno
+    try:
+        data = (<Session>deref(abstract))._recv_callback(fd, length, flags)
+    except Exception as e:
+        return -e.errno
+    cdef char* buf = data
+    memcpy(buffer, buf, len(data))
+    return len(data)
+
 
 cdef class Session:
 
     """LibSSH2 Session class providing session functions"""
 
     def __cinit__(self):
-        self._session = c_ssh2.libssh2_session_init()
+        self._session = c_ssh2.libssh2_session_init_ex(NULL,
```

Comment:
Yes (https://www.libssh2.org/libssh2_session_init_ex.html) this is what makes the callbacks functional; since the callbacks are passed a reference to the session object.